integrationTests/utils/loopbackAddressPool.mts

```diff
@@ -0,0 +1,360 @@
+ */
+async function releaseLock(): Promise<void> {
+	try {
+		await unlink(HARPER_LOOPBACK_POOL_LOCK_PATH);
+	} catch {
+		// Ignore errors if lock file is already gone
+	}
+}
+
+/**
+ * Executes a callback function while holding the lock. Automatically acquires
+ * and releases the lock, ensuring the lock is always released even if the callback
+ * throws an error.
+ *
+ * @param callback The async function to execute while holding the lock
+ * @returns The result of the callback function
+ */
+async function withLock<T>(callback: () => Promise<T>): Promise<T> {
+	await acquireLock();
+	try {
+		return await callback();
+	} finally {
+		await releaseLock();
+	}
+}
+
+/**
+ * Reads the loopback pool from the pool file. The pool is a JSON array where each
+ * index represents a loopback address (127.0.0.1, 127.0.0.2, etc.) and the value
+ * is either null (available) or a process PID (in use).
+ *
+ * If the file doesn't exist, creates and returns a new empty pool with all addresses
+ * marked as available (null).
+ *
+ * @returns The loopback pool array
+ */
+async function readPoolFile(): Promise<LoopbackPool> {
+	try {
+		const content = await readFile(HARPER_LOOPBACK_POOL_PATH, 'utf-8');
+		return JSON.parse(content) as LoopbackPool;
+	} catch (error) {
+		if ((error as NodeJS.ErrnoException).code === 'ENOENT') {
+			// If the pool file doesn't exist yet, create it with null entries
+			return Array<number | null>(HARPER_LOOPBACK_POOL_COUNT).fill(null);
+		}
+		throw error;
+	}
+}
+
+/**
+ * Writes the loopback pool to the pool file as JSON.
+ *
+ * @param pool The loopback pool array to persist
+ */
+async function writePoolFile(pool: LoopbackPool): Promise<void> {
+	await writeFile(HARPER_LOOPBACK_POOL_PATH, JSON.stringify(pool));
+}
+
+/**
+ * Finds the first available (null) index in the pool. This implements a simple
+ * first-available allocation strategy.
+ *
+ * @param pool The loopback pool array
+ * @returns The first available index, or null if the pool is full
+ */
+function findAvailableIndex(pool: LoopbackPool): number | null {
+	for (let i = 0; i < pool.length; i++) {
+		if (pool[i] === null) {
+			return i;
+		}
+	}
+	return null;
+}
+
+/**
+ * Validates the format of a loopback address and extracts the pool index.
+ *
+ * Expects addresses in the format "127.0.0.X" where X is between 1 and the pool count.
+ * The returned index is 0-based (e.g., "127.0.0.1" returns index 0).
+ *
+ * @param address The loopback address to parse (e.g., "127.0.0.2")
+ * @returns The 0-based pool index for this address
+ * @throws {InvalidLoopbackAddressError} If the address format is invalid or out of range
+ */
+function parseLoopbackAddress(address: string): number {
+	const parts = address.split('.');
+	if (parts.length !== 4 || parts[0] !== '127' || parts[1] !== '0' || parts[2] !== '0') {
+		throw new InvalidLoopbackAddressError(address);
+	}
+
+	const index = parseInt(parts[3], 10) - 1;
```

Comment:
You could use a regex instead of splitting. 

```suggestion
	const m = address.match(/^127\.0\.0\.(\d{1,3})$/);
	if (!m?.[1]) {
		throw new InvalidLoopbackAddressError(address);
	}

	const index = Number.parseInt(m[1], 10) - 1;
```

Reply:
Personally, I'm not a huge fan of using regex due to its readability. I understand it can be better performance but since this is just a util script I don't think I'm going to make this change.

Reply:
Heh, I was thinking the regex was so simple that it would be _more_ readable than the split version.